tasm-lib/src/list/higher_order/map.rs

```diff
@@ -1,330 +1,346 @@
+        fn rust_shadow(
+            &self,
+            stack: &mut Vec<BFieldElement>,
+            memory: &mut HashMap<BFieldElement, BFieldElement>,
+        ) {
+            let input_type = self.f.domain();
+            let output_type = self.f.range();
+
+            New::new(output_type.clone()).rust_shadowing(stack, vec![], vec![], memory);
+            let output_list_pointer = stack.pop().unwrap();
+
+            let input_list_pointers = (0..NUM_INPUT_LISTS)
+                .map(|_| stack.pop().unwrap())
+                .collect_vec();
+
+            // the inner function _must not_ rely on these elements
+            let chain_map_buffer_size = 3 + NUM_INPUT_LISTS;
+            let buffer = (0..chain_map_buffer_size).map(|_| random::<BFieldElement>());
+            stack.extend(buffer);
+
+            let mut total_output_len = 0;
+            for input_list_pointer in input_list_pointers {
+                let input_list_len = list_get_length(input_list_pointer, memory);
+
+                for i in (0..input_list_len).rev() {
+                    let elem = list_get(input_list_pointer, i, memory, input_type.stack_size());
+                    stack.extend(elem.into_iter().rev());
+                    self.f.apply(stack, memory);
+                    let elem = (0..output_type.stack_size())
+                        .map(|_| stack.pop().unwrap())
+                        .collect();
+                    list_set(output_list_pointer, total_output_len + i, elem, memory);
+                }
+
+                total_output_len += input_list_len;
+            }
+
+            for _ in 0..chain_map_buffer_size {
+                stack.pop();
+            }
+
+            stack.push(output_list_pointer);
+            list_set_length(output_list_pointer, total_output_len, memory);
+        }
+
+        fn pseudorandom_initial_state(
+            &self,
+            seed: [u8; 32],
+            _: Option<BenchmarkCase>,
+        ) -> FunctionInitialState {
+            let input_type = self.f.domain();
+            let mut rng = StdRng::from_seed(seed);
+            let mut stack = empty_stack();
+            let mut memory = HashMap::default();
+
+            // Some inner functions rely on additional arguments on the stack.
+            // Some tested functions panic if these arguments are too large. 😵
+            let environment_args = rng
+                .gen::<[u32; OpStackElement::COUNT]>()
+                .map(BFieldElement::from);
+            stack.extend(environment_args);
+
+            for list_length in rng.gen::<[u8; NUM_INPUT_LISTS]>() {
+                let list_length = usize::from(list_length);
+                let list_pointer = dynamic_allocator(&mut memory);
+                insert_random_list(&input_type, list_pointer, list_length, &mut memory);
+                stack.push(list_pointer);
+            }
+
+            FunctionInitialState { stack, memory }
+        }
+    }
```

Comment:
I like to create `init_state` methods on the snippet structs that capture the variables for the snippet's initial condition. In this case, such a method could be:

```rust
    impl<const NUM_INPUT_LISTS: usize> ChainMap<NUM_INPUT_LISTS> {
        fn init_state(
            &self,
            environment_args: Vec<BFieldElement>,
            list_lengths: [u16; NUM_INPUT_LISTS],
        ) -> FunctionInitialState {
            let input_type = self.f.domain();
            let mut stack = self.init_stack_for_isolated_run();
            let mut memory = HashMap::default();

            stack.extend(environment_args);

            for list_length in list_lengths {
                let list_length = usize::from(list_length);
                let list_pointer = dynamic_allocator(&mut memory);
                insert_random_list(&input_type, list_pointer, list_length, &mut memory);
                stack.push(list_pointer);
            }

            FunctionInitialState { stack, memory }
        }
    }
```

And `pseudorandom_initial_state` becomes:
```rust
        fn pseudorandom_initial_state(
            &self,
            seed: [u8; 32],
            _: Option<BenchmarkCase>,
        ) -> FunctionInitialState {
            let mut rng = StdRng::from_seed(seed);

            let environment_args = rng
                .gen::<[BFieldElement; OpStackElement::COUNT]>()
                .map(BFieldElement::from);

            let list_lengths = rng.gen::<[u8; NUM_INPUT_LISTS]>();
            let list_lengths = list_lengths.map(|x| x.into());

            self.init_state(environment_args.to_vec(), list_lengths)
        }
```

With this change, the failing test case (that required `u32` values in the "environment_args") can be rewritten as:

```rust
    #[test]
    fn test_u32_list_to_u128_list_plus_x() {
        // this code only works with 1 input list
        let raw_code = InnerFunction::RawCode(u32_to_u128_add_another_u128());
        let snippet = Map::new(raw_code);
        let encoded_u128 = random::<u128>().encode();
        let init_state = snippet.init_state(encoded_u128, [thread_rng().gen_range(0u16..200)]);
        test_rust_equivalence_given_execution_state(
            &ShadowedFunction::new(snippet),
            init_state.into(),
        );
    }
```

And with this `init_state` we get the added benefit that it becomes easy to add negative tests with the `negative_test` test helper.

Reply:
This feels a lot better. :ok_hand: